plumbing/reference_validation.go

```diff
@@ -0,0 +1,311 @@
+		}
+		break
+	case SANITIZE:
+		v.Name = ReferenceName(PatternDoubleDots.ReplaceAllString(v.Name.String(), ""))
+	}
+	return nil
+}
+
+func (v *RefNameChecker) HandleExcludedCharacters() error {
+	switch v.ActionOptions.HandleExcludedCharacters {
+	case VALIDATE:
+		if PatternExcludedCharacters.MatchString(v.Name.String()) {
+			return ErrRefExcludedCharacters
+		}
+		break
+	case SANITIZE:
+		if v.CheckRefOptions.RefSpecPattern && PatternOneAllowedAsterisk.MatchString(v.Name.String()) {
+			v.Name = ReferenceName(PatternExcludedCharactersAlternate.ReplaceAllString(v.Name.String(), ""))
+
+		} else {
+			v.Name = ReferenceName(PatternExcludedCharacters.ReplaceAllString(v.Name.String(), ""))
+		}
+	}
+	return nil
+}
+
+func (v *RefNameChecker) HandleLeadingForwardSlash() error {
+	switch v.ActionOptions.HandleLeadingForwardSlash {
+	case VALIDATE:
+		if PatternLeadingForwardSlash.MatchString(v.Name.String()) {
+			return ErrRefLeadingForwardSlash
+		}
+		break
+	case SANITIZE:
+		v.Name = ReferenceName(PatternLeadingForwardSlash.ReplaceAllString(v.Name.String(), ""))
+	}
+	return nil
+}
+
+func (v *RefNameChecker) HandleTrailingForwardSlash() error {
+	switch v.ActionOptions.HandleTrailingForwardSlash {
+	case VALIDATE:
+		if PatternTrailingForwardSlash.MatchString(v.Name.String()) {
+			return ErrRefTrailingForwardSlash
+		}
+		break
+	case SANITIZE:
+		v.Name = ReferenceName(PatternTrailingForwardSlash.ReplaceAllString(v.Name.String(), ""))
+	}
+	return nil
+}
+
+func (v *RefNameChecker) HandleConsecutiveForwardSlashes() error {
+	if SKIP == v.ActionOptions.HandleConsecutiveForwardSlashes {
+		return nil
+	}
+	if PatternConsecutiveForwardSlashes.MatchString(v.Name.String()) {
+		if SANITIZE == v.ActionOptions.HandleConsecutiveForwardSlashes {
+			if v.CheckRefOptions.Normalize {
+				v.Name = ReferenceName(PatternConsecutiveForwardSlashes.ReplaceAllString(v.Name.String(), "/"))
+				return nil
+			}
+		}
+		return ErrRefConsecutiveForwardSlashes
+	}
+	return nil
+}
+
+func (v *RefNameChecker) HandleTrailingDot() error {
+	switch v.ActionOptions.HandleTrailingDot {
+	case VALIDATE:
+		if PatternTrailingDot.MatchString(v.Name.String()) {
+			return ErrRefTrailingDot
+		}
+		break
+	case SANITIZE:
+		v.Name = ReferenceName(PatternTrailingDot.ReplaceAllString(v.Name.String(), ""))
+	}
+	return nil
+}
+
+func (v *RefNameChecker) HandleAtOpenBrace() error {
+	switch v.ActionOptions.HandleAtOpenBrace {
+	case VALIDATE:
+		if PatternAtOpenBrace.MatchString(v.Name.String()) {
+			return ErrRefAtOpenBrace
+		}
+		break
+	case SANITIZE:
+		v.Name = ReferenceName(PatternAtOpenBrace.ReplaceAllString(v.Name.String(), ""))
+	}
+	return nil
+}
+
+func (v *RefNameChecker) HandleOnlyAtSign() error {
+	if SKIP == v.ActionOptions.HandleOnlyAtSign {
+		return nil
+	}
+	if PatternOnlyAtSign.MatchString(v.Name.String()) {
+		return ErrRefOnlyAtSign
+	}
+	return nil
+}
+
+// This runs everything
+func (v *RefNameChecker) CheckRefName() error {
```

Comment:
Instead of having the ref to validate on the receiver, send it to this function as argument, also rename the function to `Validate(ref ReferenceName)`